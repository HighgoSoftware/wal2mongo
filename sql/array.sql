\set VERBOSITY terse

-- predictability
SET synchronous_commit = on;

SELECT 'init' FROM pg_create_logical_replication_slot('regression_slot', 'wal2mongo');

-- create table with different numeric type
CREATE TABLE tbl_array(id serial primary key, a bool[], b bytea[], c char[], d name[], e int2[], f int4[], g text[], h varchar[], i int8[], j float4[], k float8[], l timestamptz[], m numeric[], n uuid[] );

-- different data types for array
INSERT INTO tbl_array (a, b, c, d, e, f, g, h, i, j, k, l, m, n) VALUES(
ARRAY[true, false], 
ARRAY['\xDEADBEEF'::bytea, '\xC001CAFE'::bytea], 
ARRAY['c'::char,'h'::char], 
ARRAY['student'::name, 'teacher'::name], 
ARRAY['123'::int2, '456'::int2],
ARRAY['123456789'::int4,'987654321'::int4],
ARRAY['abc'::text, '123'::text],
ARRAY['ABCD'::varchar, '1234'::varchar],
ARRAY['112233445566778899'::int8, '998877665544332211'::int8],
ARRAY['123.456'::float4, '2222.3333'::float4],
ARRAY['123456.123'::float8, '654321.123'::float8],
ARRAY['2020-03-30 10:18:40.12-07'::timestamptz, '2020-03-30 20:28:40.12-07'::timestamptz],
ARRAY['123456789'::numeric, '987654321'::numeric],
ARRAY['40e6215d-b5c6-4896-987c-f30f3678f608'::uuid, '3f333df6-90a4-4fda-8dd3-9485d27cee36'::uuid]
);
UPDATE tbl_array SET
a=ARRAY[false, true], 
b=ARRAY['\xC001CAFE'::bytea, '\xDEADBEEF'::bytea], 
c=ARRAY['h'::char, 'c'::char], 
d=ARRAY['teacher'::name, 'student'::name], 
e=ARRAY['456'::int2, '123'::int2],
f=ARRAY['987654321'::int4, '123456789'::int4],
g=ARRAY['123'::text, 'abc'::text],
h=ARRAY['1234'::varchar, 'ABCD'::varchar],
i=ARRAY['998877665544332211'::int8, '112233445566778899'::int8],
j=ARRAY['2222.3333'::float4, '123.456'::float4],
k=ARRAY['654321.123'::float8, '123456.123'::float8],
l=ARRAY['2020-03-30 20:28:40.12-07'::timestamptz, '2020-03-30 10:18:40.12-07'::timestamptz],
m=ARRAY['987654321'::numeric, '123456789'::numeric],
n=ARRAY['3f333df6-90a4-4fda-8dd3-9485d27cee36'::uuid, '40e6215d-b5c6-4896-987c-f30f3678f608'::uuid]
WHERE id=1;
DELETE FROM tbl_array WHERE id = 1;
TRUNCATE TABLE tbl_array;


-- peek changes according to action configuration
SELECT data FROM pg_logical_slot_peek_changes('regression_slot', NULL, NULL, 'regress', 'true');

-- get changes according to action configuration
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL, 'regress', 'true');

-- drop table
DROP TABLE tbl_array;
SELECT 'end' FROM pg_drop_replication_slot('regression_slot');
