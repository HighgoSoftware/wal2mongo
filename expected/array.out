\set VERBOSITY terse
-- predictability
SET synchronous_commit = on;
SELECT 'init' FROM pg_create_logical_replication_slot('regression_slot', 'wal2mongo');
 ?column? 
----------
 init
(1 row)

-- create table with different numeric type
CREATE TABLE tbl_array(id serial primary key, a bool[], c char[], d name[], e int2[], f int4[], g text[], h varchar[], i int8[], j float4[], k float8[], l timestamptz[], m numeric[], n uuid[] );
-- different data types for array
INSERT INTO tbl_array (a, c, d, e, f, g, h, i, j, k, l, m, n) VALUES(
ARRAY[true, false], 
ARRAY['c'::char,'h'::char], 
ARRAY['student'::name, 'teacher'::name], 
ARRAY['123'::int2, '456'::int2],
ARRAY['123456789'::int4,'987654321'::int4],
ARRAY['abc'::text, '123'::text],
ARRAY['ABCD'::varchar, '1234'::varchar],
ARRAY['112233445566778899'::int8, '998877665544332211'::int8],
ARRAY['123.456'::float4, '2222.3333'::float4],
ARRAY['123456.123'::float8, '654321.123'::float8],
ARRAY['2020-03-30 10:18:40.12-07'::timestamptz, '2020-03-30 20:28:40.12-07'::timestamptz],
ARRAY['123456789'::numeric, '987654321'::numeric],
ARRAY['40e6215d-b5c6-4896-987c-f30f3678f608'::uuid, '3f333df6-90a4-4fda-8dd3-9485d27cee36'::uuid]
);
UPDATE tbl_array SET
a=ARRAY[false, true], 
c=ARRAY['h'::char, 'c'::char], 
d=ARRAY['teacher'::name, 'student'::name], 
e=ARRAY['456'::int2, '123'::int2],
f=ARRAY['987654321'::int4, '123456789'::int4],
g=ARRAY['123'::text, 'abc'::text],
h=ARRAY['1234'::varchar, 'ABCD'::varchar],
i=ARRAY['998877665544332211'::int8, '112233445566778899'::int8],
j=ARRAY['2222.3333'::float4, '123.456'::float4],
k=ARRAY['654321.123'::float8, '123456.123'::float8],
l=ARRAY['2020-03-30 20:28:40.12-07'::timestamptz, '2020-03-30 10:18:40.12-07'::timestamptz],
m=ARRAY['987654321'::numeric, '123456789'::numeric],
n=ARRAY['3f333df6-90a4-4fda-8dd3-9485d27cee36'::uuid, '40e6215d-b5c6-4896-987c-f30f3678f608'::uuid]
WHERE id=1;
DELETE FROM tbl_array WHERE id = 1;
TRUNCATE TABLE tbl_array;
-- peek changes according to action configuration
SELECT data FROM pg_logical_slot_peek_changes('regression_slot', NULL, NULL, 'regress', 'true');
                                                                                                                                                                                                                                                                                                                                     data                                                                                                                                                                                                                                                                                                                                     
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 use mycluster_mydb_regression_slot;
 db.tbl_array.insertOne( { id: NumberInt("1"), a:[true,false], c:["c","h"], d:["student","teacher"], e:[NumberInt(123),NumberInt(456)], f:[NumberInt(123456789),NumberInt(987654321)], g:["abc","123"], h:["ABCD","1234"], i:[NumberLong(112233445566778899),NumberLong(998877665544332211)], j:[123.456,2222.3333], k:[NumberDecimal("123456.123"),NumberDecimal("654321.123")], l:[ISODate("Mon Mar 30 10:18:40.12 2020 PDT"),ISODate("Mon Mar 30 20:28:40.12 2020 PDT")], m:[NumberDecimal("123456789"),NumberDecimal("987654321")], n:[UUID("40e6215d-b5c6-4896-987c-f30f3678f608"),UUID("3f333df6-90a4-4fda-8dd3-9485d27cee36")] } );
 use mycluster_mydb_regression_slot;
 db.tbl_array.updateOne( { id: NumberInt("1") }, { $set:  { id: NumberInt("1"), a:[false,true], c:["h","c"], d:["teacher","student"], e:[NumberInt(456),NumberInt(123)], f:[NumberInt(987654321),NumberInt(123456789)], g:["123","abc"], h:["1234","ABCD"], i:[NumberLong(998877665544332211),NumberLong(112233445566778899)], j:[2222.3333,123.456], k:[NumberDecimal("654321.123"),NumberDecimal("123456.123")], l:[ISODate("Mon Mar 30 20:28:40.12 2020 PDT"),ISODate("Mon Mar 30 10:18:40.12 2020 PDT")], m:[NumberDecimal("987654321"),NumberDecimal("123456789")], n:[UUID("3f333df6-90a4-4fda-8dd3-9485d27cee36"),UUID("40e6215d-b5c6-4896-987c-f30f3678f608")] } } );
 use mycluster_mydb_regression_slot;
 db.tbl_array.deleteOne( { id: NumberInt("1") } );
(6 rows)

-- get changes according to action configuration
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL, 'regress', 'true');
                                                                                                                                                                                                                                                                                                                                     data                                                                                                                                                                                                                                                                                                                                     
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 use mycluster_mydb_regression_slot;
 db.tbl_array.insertOne( { id: NumberInt("1"), a:[true,false], c:["c","h"], d:["student","teacher"], e:[NumberInt(123),NumberInt(456)], f:[NumberInt(123456789),NumberInt(987654321)], g:["abc","123"], h:["ABCD","1234"], i:[NumberLong(112233445566778899),NumberLong(998877665544332211)], j:[123.456,2222.3333], k:[NumberDecimal("123456.123"),NumberDecimal("654321.123")], l:[ISODate("Mon Mar 30 10:18:40.12 2020 PDT"),ISODate("Mon Mar 30 20:28:40.12 2020 PDT")], m:[NumberDecimal("123456789"),NumberDecimal("987654321")], n:[UUID("40e6215d-b5c6-4896-987c-f30f3678f608"),UUID("3f333df6-90a4-4fda-8dd3-9485d27cee36")] } );
 use mycluster_mydb_regression_slot;
 db.tbl_array.updateOne( { id: NumberInt("1") }, { $set:  { id: NumberInt("1"), a:[false,true], c:["h","c"], d:["teacher","student"], e:[NumberInt(456),NumberInt(123)], f:[NumberInt(987654321),NumberInt(123456789)], g:["123","abc"], h:["1234","ABCD"], i:[NumberLong(998877665544332211),NumberLong(112233445566778899)], j:[2222.3333,123.456], k:[NumberDecimal("654321.123"),NumberDecimal("123456.123")], l:[ISODate("Mon Mar 30 20:28:40.12 2020 PDT"),ISODate("Mon Mar 30 10:18:40.12 2020 PDT")], m:[NumberDecimal("987654321"),NumberDecimal("123456789")], n:[UUID("3f333df6-90a4-4fda-8dd3-9485d27cee36"),UUID("40e6215d-b5c6-4896-987c-f30f3678f608")] } } );
 use mycluster_mydb_regression_slot;
 db.tbl_array.deleteOne( { id: NumberInt("1") } );
(6 rows)

-- drop table
DROP TABLE tbl_array;
SELECT 'end' FROM pg_drop_replication_slot('regression_slot');
 ?column? 
----------
 end
(1 row)

